core/src/main/java/org/apache/accumulo/fate/zookeeper/DistributedReadWriteLock.java

```diff
@@ -218,22 +228,52 @@ public boolean tryLock() {
       }
       SortedMap<Long,byte[]> entries = qlock.getEarlierEntries(entry);
       Iterator<Entry<Long,byte[]>> iterator = entries.entrySet().iterator();
-      if (!iterator.hasNext())
+      if (!iterator.hasNext()) {
         throw new IllegalStateException("Did not find our own lock in the queue: " + this.entry
             + " userData " + new String(this.userData, UTF_8) + " lockType " + lockType());
-      return iterator.next().getKey().equals(entry);
+      }
+      if (iterator.next().getKey().equals(entry)) {
+        return true;
+      }
+      if (preemptBlockingTransactions) {
+        log.debug("Preempting transactions that are blocking this transaction");
+        // Loop through all of the prior transactions that are waiting to
+        // acquire this write lock. If the transaction has not succeeded or failed,
+        // then fail it and return false from this method so that Utils.reserveX()
+        // will call this method again and will re-check the prior transactions.
+        iterator = entries.entrySet().iterator();
+        while (iterator.hasNext()) {
+          Entry<Long,byte[]> e = iterator.next();
+          Long priorEntry = e.getKey();
+          ParsedLock priorLock = new ParsedLock(e.getValue());
+          long priorTxid = Long.valueOf(new String(priorLock.getUserData(), UTF_8), 16);
+          if (!priorEntry.equals(entry)) {
+            log.info("Attempting to preempt blocking tx: {}", Long.toHexString(priorTxid));
+            store.cancel(priorTxid);
```

Comment:
If its not currently reserved, could possibly reserved it and change the state to FAILED_IN_PROGRESS.

```java
  if(reserved)
     interrupt()
  else
    changeToFailedInProgress()
```

Reply:
Addressed in d5df49354bf83d9174445c5e669936bb5cdf473e